upup/pkg/kutil/cluster_resources.go

```diff
@@ -0,0 +1,170 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kutil
+
+import (
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kops/upup/pkg/fi/cloudup/awsup"
+	"k8s.io/kubernetes/pkg/util/sets"
+	"strings"
+)
+
+// Todo
+// After this is stable, we should pull this out of kutil so we can start mapping out the implementation
+// for each cloud. Just taking the first steps now into separating these out, but we probably have some
+// bigger concerns as this grows.. Just starting to draw some boundaries.
+
+// ClusterResources is a representation of a cluster with abilities to ListResources and DeleteResources
+type ClusterResources interface {
+	ListResources() (map[string]*ResourceTracker, error)
+	DeleteResources(resources map[string]*ResourceTracker) error
+}
+
+
+// AwsCluster is an implementation of ClusterResources
+// The algorithm is pretty simple: it discovers all the resources it can (primary using tags)
+// There are a few tweaks to that approach, like choosing a default ordering, but it is not much
+// smarter.
+// Some dependencies are invisible (e.g. ELB dependencies).
+//
+type AwsCluster struct {
+	ClusterName string
+	Cloud       fi.Cloud
+}
+
+func (c *AwsCluster) ListResources() (map[string]*ResourceTracker, error) {
+	cloud := c.Cloud.(awsup.AWSCloud)
+
+	resources := make(map[string]*ResourceTracker)
+
+	// These are the functions that are used for looking up
+	// cluster resources by their tags.
+	listFunctions := []listFn{
+		// EC2
+		ListInstances,
```

Comment:
Trying not to change delete logic in this PR - just change the way we interface with it..